numbat/src/typechecker.rs

```diff
@@ -661,10 +697,35 @@ impl TypeChecker {
             ));
         }
 
-        let return_type = match return_type {
-            Type::Dimension(d) => Type::Dimension(substitute(&substitutions, d)),
-            type_ => type_.clone(),
-        };
+        fn apply_substitutions(
+            t: &Type,
+            substitute: impl Fn(&[(String, DType)], &DType) -> DType + Copy,
+            substitutions: &[(String, DType)],
+        ) -> Type {
+            match t {
+                Type::Dimension(d) => Type::Dimension(substitute(&substitutions, d)),
+                Type::Struct(StructInfo {
+                    definition_span,
+                    name,
+                    fields,
+                }) => Type::Struct(StructInfo {
+                    definition_span: *definition_span,
+                    name: name.clone(),
+                    fields: fields
+                        .into_iter()
+                        .map(|(n, (s, t))| {
+                            (
+                                n.to_owned(),
+                                (s.clone(), apply_substitutions(t, substitute, substitutions)),
+                            )
+                        })
+                        .collect(),
+                }),
+                type_ => type_.clone(),
```

Comment:
Can this code even trigger? We can't have structs with generic parameters, can we?